lib/lndk.js

```diff
@@ -0,0 +1,249 @@
+  4: { bit: 4, type: 'UPFRONT_SHUTDOWN_SCRIPT_REQ', is_required: true },
+  5: { bit: 5, type: 'UPFRONT_SHUTDOWN_SCRIPT_OPT', is_required: false },
+  6: { bit: 6, type: 'GOSSIP_QUERIES_REQ', is_required: true },
+  7: { bit: 7, type: 'GOSSIP_QUERIES_OPT', is_required: false },
+  8: { bit: 8, type: 'TLV_ONION_REQ', is_required: true },
+  9: { bit: 9, type: 'TLV_ONION_OPT', is_required: false },
+  10: { bit: 10, type: 'EXT_GOSSIP_QUERIES_REQ', is_required: true },
+  11: { bit: 11, type: 'EXT_GOSSIP_QUERIES_OPT', is_required: false },
+  12: { bit: 12, type: 'STATIC_REMOTE_KEY_REQ', is_required: true },
+  13: { bit: 13, type: 'STATIC_REMOTE_KEY_OPT', is_required: false },
+  14: { bit: 14, type: 'PAYMENT_ADDR_REQ', is_required: true },
+  15: { bit: 15, type: 'PAYMENT_ADDR_OPT', is_required: false },
+  16: { bit: 16, type: 'MPP_REQ', is_required: true },
+  17: { bit: 17, type: 'MPP_OPT', is_required: false },
+  18: { bit: 18, type: 'WUMBO_CHANNELS_REQ', is_required: true },
+  19: { bit: 19, type: 'WUMBO_CHANNELS_OPT', is_required: false },
+  20: { bit: 20, type: 'ANCHORS_REQ', is_required: true },
+  21: { bit: 21, type: 'ANCHORS_OPT', is_required: false },
+  22: { bit: 22, type: 'ANCHORS_ZERO_FEE_HTLC_REQ', is_required: true },
+  23: { bit: 23, type: 'ANCHORS_ZERO_FEE_HTLC_OPT', is_required: false },
+  24: { bit: 24, type: 'ROUTE_BLINDING_REQUIRED', is_required: true },
+  25: { bit: 25, type: 'ROUTE_BLINDING_OPTIONAL', is_required: false },
+  30: { bit: 30, type: 'AMP_REQ', is_required: true },
+  31: { bit: 31, type: 'AMP_OPT', is_required: false }
+}
+
+const chainsMap = {
+  '06226e46111a0b59caaf126043eb5bbf28c34f3a5e332a1fc7b2b73cf188910f': 'regtest',
+  '43497fd7f826957108f4a30fd9cec3aeba79972084e90ead01ea330900000000': 'testnet',
+  '6fe28c0ab6f1b372c1a6a246ae63f74f931e8365e15a089c68d6190000000000': 'mainnet'
+}
+
+export async function parseBolt12Request ({
+  lnd,
+  request
+}) {
+  const lndk = lnd?.lndk
+  if (!lndk) throw new Error('lndk not installed, please use installLNDK')
+
+  const invoice_hex_str = request.startsWith('lni1') ? bech32b12.decode(request.slice(4)).toString('hex') : request
+
+  const invoice_contents = await new Promise((resolve, reject) => {
+    lndk.DecodeInvoice({
+      invoice: invoice_hex_str
+    }, (error, response) => {
+      if (error) return reject(error)
+      resolve(response)
+    })
+  })
+
+  const {
+    amount_msats,
+    description,
+    node_id,
+    chain,
+    payment_hash,
+    created_at,
+    relative_expiry,
+    features,
+    payer_note
+  } = invoice_contents
+
+  // convert from lndk response to ln-service parsePaymentRequest output layout
+  let minCltvDelta
+  for (const path of invoice_contents.payment_paths) {
+    const info = path.blinded_pay_info
+    if (minCltvDelta === undefined || info.cltv_expiry_delta < minCltvDelta) {
+      minCltvDelta = info.cltv_expiry_delta
+    }
+  }
+
+  const out = {
```

Comment:
This function converts the output from lndk to an object with a layout compatible with the output from ln-service.

I've followed the [doc](https://github.com/alexbosworth/ln-service/blob/158597a28c3b392b98a175801ad56ade535b3195/README.md?plain=1#L939) so it should match the output very closely, but honestly i don't like the fact that ln-service returns  so many values that mean the same thing with different types (eg, tokens, mtokens, safe_tokens...), i've implemented this conversion because it makes our code simpler with fewer changes, since we can handle both bolt11 and bolt12 invoices in the same way, but imo we should provide our own abstraction at some point, using only bigint msats and things that make more sense for our codebase.

Reply:
I agree, ln-service tries to make the grpc interface nicer, but it isn't as well documented as the grpc interface so it ends up just being confusing most of the time.